In [1]:
import numpy as np
import itertools
import math

### Даны три множества a,b и с. Необходимо выполнить все изученные виды бинарных операций над всеми комбинациями множеств.

In [2]:
A = {1, 3, 5, 7, 9}
B = {1, 2, 3, 4, 5}
C = {5, 6, 7, 8, 9}

### Пересечение

In [3]:
set.intersection(A, B)

{1, 3, 5}

In [4]:
set.intersection(B, C)

{5}

In [5]:
set.intersection(C, A)

{5, 7, 9}

### Объединение

In [6]:
set.union(A, B)

{1, 2, 3, 4, 5, 7, 9}

In [7]:
set.union(B, C)

{1, 2, 3, 4, 5, 6, 7, 8, 9}

In [8]:
set.union(C, A)

{1, 3, 5, 6, 7, 8, 9}

### Разность

In [9]:
set.difference(A, B)

{7, 9}

In [10]:
set.difference(B, A)

{2, 4}

In [11]:
set.difference(B, C)

{1, 2, 3, 4}

In [12]:
set.difference(C, B)

{6, 7, 8, 9}

In [13]:
set.difference(C, A)

{6, 8}

In [14]:
set.difference(A, C)

{1, 3}

### Симметрическая разность

In [15]:
set.symmetric_difference(A, B)

{2, 4, 7, 9}

In [16]:
set.symmetric_difference(B, C)

{1, 2, 3, 4, 6, 7, 8, 9}

In [17]:
set.symmetric_difference(C, A)

{1, 3, 6, 8}

### Декартово произведение

In [18]:
tmp = set()
for i in itertools.product(A,B):
    tmp.add(i)
print(tmp)

{(7, 3), (1, 3), (9, 1), (9, 4), (5, 1), (7, 2), (1, 2), (3, 3), (5, 5), (1, 5), (3, 4), (1, 1), (3, 2), (5, 4), (7, 1), (9, 3), (1, 4), (7, 5), (5, 3), (3, 5), (9, 2), (3, 1), (7, 4), (9, 5), (5, 2)}


In [19]:
tmp = set()
for i in itertools.product(B,C):
    tmp.add(i)
print(tmp)

{(5, 9), (4, 7), (4, 8), (5, 6), (2, 8), (1, 6), (3, 7), (2, 5), (5, 8), (4, 9), (5, 5), (2, 9), (1, 5), (3, 6), (2, 6), (4, 5), (3, 9), (1, 9), (3, 5), (2, 7), (4, 6), (5, 7), (3, 8), (1, 8), (1, 7)}


In [20]:
tmp = set()
for i in itertools.product(C,A):
    tmp.add(i)
print(tmp)

{(5, 9), (6, 9), (7, 3), (9, 1), (7, 7), (8, 9), (5, 1), (8, 5), (6, 7), (5, 5), (8, 1), (6, 3), (9, 7), (7, 1), (9, 3), (7, 5), (8, 7), (6, 5), (7, 9), (5, 3), (8, 3), (6, 1), (5, 7), (9, 9), (9, 5)}


### *На языке Python предложить алгоритм вычисляющий численно предел с точностью e = 10^-7

In [21]:
n = 1
f_n_prev = 0
epsilon = 1e-9
while True:
    try:
        f_n = n / (math.factorial(n) ** (1/n))
    except OverflowError:
        print(n)
        break
    if abs(f_n_prev - f_n) < epsilon:
        break
    f_n_prev = f_n
    n += 1
print(f_n)

171
2.663087878748024


### Расчет "в лоб" очень быстро приводик к OverflowError (на 171 элементе), поэтому заменяем n! по формуле Муавра — Стирлинга n!~(math.sqrt(2 * math.pi * n)*(n/e)^n)

In [22]:
n = 1 # Начальный номер
epsilon = 1e-7 # Эпсилон, погрешность (здесь будем считать разность между элементом n и 10*n)
while True:
    # Расчет для элемента n
    f_n = 1 / (math.sqrt(2 * math.pi * n)**(1/n) * (1 / math.e ))
    # Расчет для элемента 10*n
    f_n_k = 1 / (math.sqrt(2 * math.pi * n*10)**(1/(n*10)) * (1 / math.e ))
    dif = abs(f_n - f_n_k) # Погрешность для данного n
    if dif < epsilon:
        break
    n += (1 + int(1/dif)) # int(1/dif) - используем для ускорения процесса
print(f'предел, lim = {f_n}')
print(f'погрешность = {dif}')
print(f'n = {n}')

предел, lim = 2.7182817161016715
погрешность = 9.990131344395081e-08
n = 256451635
